### PR TITLE
Create Payment Requests

### PR DESCRIPTION
**Pull request by ghickman**
This adds the Request Payment modal/form, moves the funding blocks into the Project detail page, and populates those blocks with the relevant information from PaymentRequests.

### CODE REVIEW

**Comment by ghickman** [comment]
The form UI bugs from #1411 mean I'm not able to test the MultiFieldField properly.  Should we gate this PR on that or revist after that's fixed?

File: opentech/apply/projects/templates/application_projects/includes/payment_requests.html [hunk 1]
```diff
@@ -10,44 +15,45 @@
...
-                <th class="payment-block__table-update">Status</th>
+                <th class="payment-block__table-update">Change status</th>
             </tr>
         </thead>
         <tbody>
+            {% for payment_request in payments.not_rejected %}
             <tr>
-                <td><span class="payment-block__mobile-label">Amount: </span>$10,000</td>
+                <td><span class="payment-block__mobile-label">Amount: </span>${{ payment_request.value }}</td>
                 <td><span class="payment-block__mobile-label">Status: </span>Approved 1111-11-11</td>
```
Comment (todd-dembrey) [inline opentech/apply/projects/templates/application_projects/includes/payment_requests.html]: `{{ payment_request.status }}`

File: opentech/apply/projects/templates/application_projects/includes/payment_requests.html [hunk 2]
```diff
@@ -10,44 +15,45 @@
...
+    </p>
 
     <table class="payment-block__table is-hidden js-payment-block-rejected-table">
         <thead>
             <tr>
                 <th class="payment-block__table-amount">Amount</th>
                 <th class="payment-block__table-status">Status</th>
                 <th class="payment-block__table-docs">Documents</th>
-                <th class="payment-block__table-update">Status</th>
+                <th class="payment-block__table-update">Change status</th>
```
Comment (todd-dembrey) [inline opentech/apply/projects/templates/application_projects/includes/payment_requests.html]: We didn't have a design for the rejected table, but i don't think it makes sense to include the status change column.

File: opentech/apply/projects/templates/application_projects/includes/payment_requests.html [hunk 3]
```diff
@@ -10,44 +15,45 @@
...
             </tr>
         </thead>
         <tbody>
+            {% for request in payments.rejected %}
             <tr>
                 <td><span class="payment-block__mobile-label">Amount: </span>$14,000</td>
                 <td><span class="payment-block__mobile-label">Status: </span>Rejected 1111-11-11</td>
```
Comment (todd-dembrey) [inline opentech/apply/projects/templates/application_projects/includes/payment_requests.html]: `{{ payment_request.value }}` 
`{{ payment_request.status }}`

File: opentech/apply/projects/views.py [hunk 4]
```diff
@@ -82,6 +90,41 @@ def get_contracts(self):
...
+            'totals': self.get_totals(project),
+        }
+
+        context = super().get_context_data(**kwargs)
+        context['payments'] = payments
+        return context
+
+    def get_totals(self, project):
+        def percentage(total, value):
+            return round((value / total) * 100)
```
Comment (todd-dembrey) [inline opentech/apply/projects/views.py]: We should be working in Decimal at this point - anything to do with money. 
`(value / total * 100).quantize(Decimal('1'), rounding=ROUND_DOWN)`

File: opentech/apply/projects/views.py [hunk 5]
```diff
@@ -82,6 +90,41 @@ def get_contracts(self):
...
+
+        unpaid_requests = project.payment_requests.filter(Q(status=SUBMITTED) | Q(status=UNDER_REVIEW))
+        awaiting_absolute = sum(unpaid_requests.values_list('value', flat=True))
```
Comment (todd-dembrey) [inline opentech/apply/projects/views.py]: Alternative?

`awaiting_absolute = unpaid_requests.aggregate(total_value=Sum('value'))['total_value']`

maybe i just like over using the ORM

File: opentech/apply/projects/tests/test_views.py [hunk 6]
```diff
@@ -527,3 +528,54 @@ def test_approve_unsigned_contract(self):
 
         messages = list(response.context['messages'])
         self.assertEqual(len(messages), 1)
+
+
+class TestRequestPaymentView(BaseViewTestCase):
+    base_view_name = 'detail'
+    url_name = 'funds:projects:{}'
+    user_factory = StaffFactory
```
Comment (todd-dembrey) [inline opentech/apply/projects/tests/test_views.py]: Same again for applicant please.

File: opentech/apply/projects/views.py (hunk 4 above)
Comment (ghickman) [reply opentech/apply/projects/views.py]: Fixed and rebased.

File: opentech/apply/projects/templates/application_projects/includes/payment_requests.html (hunk 3 above)
Comment (ghickman) [reply opentech/apply/projects/templates/application_projects/includes/payment_requests.html]: Fixed and rebased.

File: opentech/apply/projects/templates/application_projects/includes/payment_requests.html (hunk 1 above)
Comment (ghickman) [reply opentech/apply/projects/templates/application_projects/includes/payment_requests.html]: Fixed and rebased.

File: opentech/apply/projects/models.py [hunk 7]
```diff
@@ -189,6 +249,12 @@ def get_absolute_url(self):
     def can_make_approval(self):
         return self.is_locked and self.status == COMMITTED
 
+    def can_request_funding(self):
+        """
+        Should we show this Project's funding block?
+        """
+        return self.status == IN_PROGRESS
```
[the remark this answers is not in the document]
Comment (ghickman) [reply opentech/apply/projects/models.py]: Fixed and rebased.

File: opentech/apply/projects/tests/test_views.py (hunk 6 above)
Comment (ghickman) [reply opentech/apply/projects/tests/test_views.py]: Fixed and rebased.

File: opentech/apply/projects/views.py (hunk 5 above)
Comment (ghickman) [reply opentech/apply/projects/views.py]: Can do, it's the same thing but pushes the calculation to the database.  At the scale I believe this application runs at the difference is negligible.  Would you like me to change it?

File: opentech/apply/projects/models.py [hunk 8]
```diff
@@ -73,14 +81,67 @@ def get_remove_form(self):
...
+    (PAID, 'Paid'),
+    (DECLINED, 'Declined'),
+]
+
+
+class PaymentRequest(models.Model):
+    project = models.ForeignKey("Project", on_delete=models.CASCADE, related_name="payment_requests")
+    by = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE, related_name="payment_requests")
+
+    invoice = models.FileField(upload_to=invoice_path)
```
Comment (todd-dembrey) [inline opentech/apply/projects/models.py]: Needs to upload to the private storage as per #1417 - i think we need to merge sprint-1 into 2 to allow that.

**Pull request merged**